classes/plugins/PluginHelper.php

```diff
@@ -192,74 +173,86 @@ protected function _getConnectionParams()
 
-        // Copy the plug-in from the temporary folder to the target folder.
-        if (!$fileManager->copyDir($path, $pluginDest)) {
-            throw new Exception('Could not copy plugin to desination!');
-        }
-        if (!$fileManager->rmtree($path)) {
-            throw new Exception('Could not remove temporary plugin path!');
-        }
+            // Delete existing files.
+            $fileManager->rmtree($destinyPath);
 
-        $upgradeFile = $pluginDest . '/' . self::PLUGIN_UPGRADE_FILE;
-        if ($fileManager->fileExists($upgradeFile)) {
-            $siteDao = DAORegistry::getDAO('SiteDAO'); /** @var SiteDAO $siteDao */
-            $site = $siteDao->getSite();
-            $params = $this->_getConnectionParams();
-            $params['locale'] = $site->getPrimaryLocale();
-            $params['additionalLocales'] = $site->getSupportedLocales();
-            $installer = new Upgrade($params, $upgradeFile, true);
+            // Check whether deleting has worked.
+            if (is_dir($destinyPath)) {
+                throw new Exception(__('manager.plugins.deleteError', ['pluginName' => $pluginVersion->getProduct()]));
+            }
 
-            if (!$installer->execute()) {
-                throw new Exception(__('manager.plugins.upgradeFailed', ['errorString' => $installer->getErrorString()]));
+            // Copy the plug-in from the temporary folder to the target folder.
+            $fileManager->copyDir($sourcePath, $destinyPath) || throw new Exception('Could not copy plugin to destination!');
+
+            try {
+                $upgradeFile = "{$destinyPath}/" . static::PLUGIN_UPGRADE_FILE;
+                if ($fileManager->fileExists($upgradeFile)) {
+                    /** @var SiteDAO */
+                    $siteDao = DAORegistry::getDAO('SiteDAO');
+                    $site = $siteDao->getSite();
+                    $params = $this->_getConnectionParams();
+                    $params['locale'] = $site->getPrimaryLocale();
+                    $params['additionalLocales'] = $site->getSupportedLocales();
+                    $installer = new Upgrade($params, $upgradeFile, true);
+                    // Run the upgrade/migration
+                    $installer->execute() || throw new Exception(__('manager.plugins.upgradeFailed', ['errorString' => $installer->getErrorString()]));
+                }
+
+                // Add the new version to the database
+                $pluginVersion->setCurrent(1);
+                $versionDao->insertVersion($pluginVersion, true);
+                return $pluginVersion;
+            } catch (Throwable $e) {
+                // Delete the plugin files on failure
+                $fileManager->rmtree($destinyPath);
+                throw $e;
             }
+        } finally {
+            // Discard the temporary plugin files
+            $fileManager->rmtree($sourcePath);
         }
+    }
 
-        $installedPlugin->setCurrent(0);
```

Comment:
Though I don't see an update call anywhere here, I would guess this was originally to set the `versions.current` column to `0` for the old version of the plugin. Is this being done elsewhere now that the code has been rewritten?

Reply:
I suspected this line wasn't needed because there was no `$versionDao->insertVersion($installedPlugin);` after that, then I confirmed here:
https://github.com/pkp/pkp-lib/blob/87e018ebe033574fa40749f0de110c5875f2f56b/classes/site/VersionDAO.php#L116